src/ucp/proto/proto_select.c

```diff
@@ -25,6 +28,221 @@ typedef struct {
     size_t                         priv_offsets[UCP_PROTO_MAX_COUNT];
 } ucp_proto_select_init_protocols_t;
 
+/* Temporary list of constructed protocol thresholds */
+typedef struct {
+    size_t                         max_length;
+    ucp_proto_id_t                 proto_id;
+} ucp_proto_threshold_tmp_elem_t;
+
+
+UCS_ARRAY_DEFINE_INLINE(ucp_proto_thresh, unsigned,
+                        ucp_proto_threshold_tmp_elem_t);
+
+const ucp_proto_threshold_elem_t*
+ucp_proto_thresholds_search_slow(const ucp_proto_threshold_elem_t *thresholds,
+                                 size_t msg_length)
+{
+    unsigned idx;
+    for (idx = 0; msg_length > thresholds[idx].max_msg_length; ++idx);
+    return &thresholds[idx];
+}
+
+static ucs_status_t
+ucp_proto_thresholds_append(ucs_array_t(ucp_proto_thresh) *thresh_list,
+                            size_t max_length, ucp_proto_id_t proto_id)
+{
+    ucp_proto_threshold_tmp_elem_t *thresh_elem;
+    ucs_status_t status;
+
+    /* Consolidate with last protocol if possible */
+    if (!ucs_array_is_empty(thresh_list)) {
+        thresh_elem = ucs_array_last(thresh_list);
+        ucs_assertv(max_length > thresh_elem->max_length,
+                    "max_length=%zu last->max_length=%zu",
+                    max_length, thresh_elem->max_length);
+        if (thresh_elem->proto_id == proto_id) {
+            thresh_elem->max_length = max_length;
+            return UCS_OK;
+        }
+    }
+
+    status = ucs_array_append(ucp_proto_thresh, thresh_list);
+    if (status != UCS_OK) {
+        return status;
+    }
+
+    thresh_elem             = ucs_array_last(thresh_list);
+    thresh_elem->max_length = max_length;
+    thresh_elem->proto_id   = proto_id;
+    return UCS_OK;
+}
+
+static ucs_status_t
+ucp_proto_thresholds_select_best(ucp_proto_id_mask_t proto_mask,
```

Comment:
maybe check if `proto_mask` is not empty and if it is empty return `UCS_ERR_UNSUPPORTED`?

Reply:
it's already checked it caller function, will add assertion